File: src/core/param_controller.cpp
/**
 * @file param_controller.cpp
 * @brief PPG parameter controller implementation
 * @version 2.0.0
 * @date 25 April 2026
 */

#include "core/param_controller.h"

// ============================================================================
// CONSTRUCTOR
// ============================================================================
ParamController::ParamController() {
    resetToDefaults();
}

// ============================================================================
// SET CONDITION
// ============================================================================
void ParamController::setCondition(uint8_t condition) {
    PPGParameters pending = currentPPG;
    pending.condition = static_cast<PPGCondition>(condition);

    // Re-clamp parameters to new condition limits
    PPGLimits newLimits = getPPGLimits(pending.condition);
    pending.heartRate = clamp(pending.heartRate, newLimits.heartRate.min, newLimits.heartRate.max);
    pending.perfusionIndex = clamp(pending.perfusionIndex, newLimits.perfusionIndex.min, newLimits.perfusionIndex.max);
    pending.dicroticNotch = clamp(pending.dicroticNotch, newLimits.dicroticNotch.min, newLimits.dicroticNotch.max);

    pendingPPG.hasPending = true;
    pendingPPG.pendingValue = pending;
    pendingPPG.requestTime = millis();
}

// ============================================================================
// TYPE A PARAMETERS (IMMEDIATE)
// ============================================================================
void ParamController::setNoiseLevel(float noise) {
    noise = clamp(noise, 0.0f, 0.10f);
    currentPPG.noiseLevel = noise;
}

void ParamController::setAmplitude(float amplitude) {
    float ampFactor = clamp(amplitude / 100.0f, 0.5f, 2.0f);
    currentPPG.amplification = ampFactor;
}

// ============================================================================
// TYPE B PARAMETERS (DEFERRED)
// ============================================================================
void ParamController::setHeartRate(float hr) {
    PPGLimits limits = getPPGLimits(currentPPG.condition);
    PPGParameters pending = currentPPG;
    pending.heartRate = clamp(hr, limits.heartRate.min, limits.heartRate.max);
    pendingPPG.hasPending = true;
    pendingPPG.pendingValue = pending;
    pendingPPG.requestTime = millis();
}

void ParamController::setPerfusionIndex(float pi) {
    PPGLimits limits = getPPGLimits(currentPPG.condition);
    PPGParameters pending = currentPPG;
    pending.perfusionIndex = clamp(pi, limits.perfusionIndex.min, limits.perfusionIndex.max);
    pendingPPG.hasPending = true;
    pendingPPG.pendingValue = pending;
    pendingPPG.requestTime = millis();
}

// ============================================================================
// APPLY PENDING PARAMETERS
// ============================================================================
bool ParamController::applyPendingParams() {
    if (pendingPPG.hasPending) {
        currentPPG = pendingPPG.pendingValue;
        pendingPPG.hasPending = false;
        return true;
    }
    return false;
}

bool ParamController::hasPendingParams() const {
    return pendingPPG.hasPending;
}

// ============================================================================
// GETTERS
// ============================================================================
PPGLimits ParamController::getCurrentLimits() const {
    return getPPGLimits(currentPPG.condition);
}

// ============================================================================
// VALIDATION
// ============================================================================
bool ParamController::validateParams(const PPGParameters& params) {
    PPGLimits limits = getPPGLimits(params.condition);
    if (params.heartRate < limits.heartRate.min || params.heartRate > limits.heartRate.max) return false;
    if (params.noiseLevel < 0.0f || params.noiseLevel > 0.10f) return false;
    if (params.perfusionIndex < limits.perfusionIndex.min || params.perfusionIndex > limits.perfusionIndex.max) return false;
    if (params.dicroticNotch < limits.dicroticNotch.min || params.dicroticNotch > limits.dicroticNotch.max) return false;
    return true;
}

// ============================================================================
// RESET TO DEFAULTS
// ============================================================================
void ParamController::resetToDefaults() {
    currentPPG = PPGParameters();
    pendingPPG.hasPending = false;
}

// ============================================================================
// UTILITY
// ============================================================================
float ParamController::clamp(float value, float min, float max) {
    if (value < min) return min;
    if (value > max) return max;
    return value;
}

File: PPG_simulator_raspi/core/param_controller.py
"""
param_controller.py — PPG parameter controller with validation and clamping.

Port of param_controller.cpp.
"""

from models.ppg_model import PPGParameters, get_ppg_limits, _clamp


class ParamController:
    """Manages PPG parameter updates with condition-specific limits."""

    def __init__(self):
        self.current = PPGParameters()
        self._pending = None
        self._has_pending = False

    def set_condition(self, condition: int):
        pending = self.current.copy()
        pending.condition = condition
        lim = get_ppg_limits(condition)
        pending.heart_rate = _clamp(pending.heart_rate, lim.heart_rate.min, lim.heart_rate.max)
        pending.perfusion_index = _clamp(pending.perfusion_index, lim.perfusion_index.min, lim.perfusion_index.max)
        pending.dicrotic_notch = _clamp(pending.dicrotic_notch, lim.dicrotic_notch.min, lim.dicrotic_notch.max)
        self._pending = pending
        self._has_pending = True

    def set_noise_level(self, noise: float):
        self.current.noise_level = _clamp(noise, 0.0, 0.10)

    def set_heart_rate(self, hr: float):
        lim = get_ppg_limits(self.current.condition)
        p = self.current.copy()
        p.heart_rate = _clamp(hr, lim.heart_rate.min, lim.heart_rate.max)
        self._pending = p; self._has_pending = True

    def set_perfusion_index(self, pi: float):
        lim = get_ppg_limits(self.current.condition)
        p = self.current.copy()
        p.perfusion_index = _clamp(pi, lim.perfusion_index.min, lim.perfusion_index.max)
        self._pending = p; self._has_pending = True

    def set_spo2(self, spo2: float):
        lim = get_ppg_limits(self.current.condition)
        p = self.current.copy()
        p.spo2 = _clamp(spo2, lim.spo2.min, lim.spo2.max)
        self._pending = p; self._has_pending = True

    def set_resp_rate(self, rr: float):
        lim = get_ppg_limits(self.current.condition)
        p = self.current.copy()
        p.resp_rate = _clamp(rr, lim.resp_rate.min, lim.resp_rate.max)
        self._pending = p; self._has_pending = True

    def apply_pending(self) -> bool:
        if self._has_pending and self._pending:
            self.current = self._pending
            self._has_pending = False
            return True
        return False

    def has_pending(self) -> bool:
        return self._has_pending

    def get_limits(self):
        return get_ppg_limits(self.current.condition)

    def validate(self, params: PPGParameters) -> bool:
        lim = get_ppg_limits(params.condition)
        if not (lim.heart_rate.min <= params.heart_rate <= lim.heart_rate.max): return False
        if not (0 <= params.noise_level <= 0.10): return False
        if not (lim.perfusion_index.min <= params.perfusion_index <= lim.perfusion_index.max): return False
        if not (lim.spo2.min <= params.spo2 <= lim.spo2.max): return False
        if not (lim.resp_rate.min <= params.resp_rate <= lim.resp_rate.max): return False
        return True

    def reset_to_defaults(self):
        self.current = PPGParameters()
        self._has_pending = False
